enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,141 @@
+However, the services on masters and workers nodes will be managed by a new operator instead of static pods.
+
+The following functionalities are expected as part of the new operator:
+* Introduce a CRD used by the operator to manage the self-hosted stack.
+* Fields in CR spec that control the self-hosted stack configuration.
+* Fields in CR status that expose self-hosted stack status (e.g: which node holds the API-VIP).
+* Run the self-hosted stack components (Keepalived, Coredns-MDNS, HAProxy, MDNS-publisher) as a K8S applications.
+
+### Kubelet and Operator circular dependency: 
+In order to run the operator's component, Kubelet in each node should first join the cluster by communicating with the control plane, however the components provided by the operator are responsible for enabling that.
+As a result, the below circular dependency should be addressed:
+* Kubelet can't talk to the control plane until the operator has started.
+* Operator can't start until Kubelet can talk to the control plane.
+
+To break the circular dependency, the operator should start running before bootstrap is completed in order to assure that bootstrap owns the API-VIP. In addition, a new /etc/hosts entry should be added to the nodes using MCO ignition.  This /etc/hosts entry is expected to resolve api-int hostname to the API-VIP address. 
+
+### CRD definition
+
+Below is a possible option for a CRD instance of this operator:
+
+
+```yaml
+apiVersion: clusterhosted.operator.io/v1alpha1
+kind: ClusterHostedConfig
+metadata:
+  name: clusterhostedconfig-sample
+spec:
+  dns: Enable
+  loadbalancer: Enable
+status:
+  apiserverinternalipowner: master-0
+  ingressipowner: worker-1
+```
+
+This manifest creates DNS and loadbalancer self-hosted stack, from the status section we can see that master-0 node currently owns api-int VIP and worker-1 node owns ingress VIP. Additional spec and status fields may be introduced based on requirements.
+
+### User Stories
+
+#### As an admin I want to disable the self-hosted on-prem LB and DNS stack.
+
+#### As an admin I want to use the self-hosted on-prem LB and DNS stack also in UPI deployments.
+
+#### As an admin I want to retrieve the self-hosted stack status through K8S API.
+
+
+
+### Upgrade / Downgrade Strategy
+
+Prior to this proposal, the self-hosted stack implementation was running through the MCO.
+The implemntation of the self-hostd stack using the new operator might lead to cases during upgrade/downgrade in which two parallel components are running the self-hosted stack. 
+
+
+It's required to verify that LB and DNS services are available in case where both MCO and the new operator run the self-hosted stack.
+
+[TODO] It's required to further understand the availability of LB and DNS services when both MCO and new operator run simultaneously. 
+
```

Comment:
The upgrade/downgrade section should be expanded to explain what the process would be.

Can we assume the nodes are updated _after_ the new operator is deployed in the cluster? Is there a risk no component is running the self-hosted stack? Does it apply to downgrade too?

How would it work when the two stacks are running at the same time? In my understanding, they shouldn't conflict if they use different IP addresses. Is the following interpretation correct?

On upgrade:
- start the new operator, serving LB and DNS on different IPs than the existing static pods
- apply the new MCO templates on the nodes, effectively removing static pods, and
  - point `api-int` to new LB address via `/etc/hosts`.
  - update `/etc/resolv.conf` to point to the new resolver

On downgrade, the order needs to be reversed:
- revert to MCO templates with static pods
- stop the operator once all nodes have reverted to old MCO template

Reply:
Yep, the upgrade/downgrade processes should be described here in more detail.
I think we should work according to the following assumptions/requirements:
* the same IP addresses (API, Ingress VIps)  should be used.
* The API ,Ingress and DNS services should be available during upgrade/downgrade time.

One direction for the upgrade path could shutting down the  MCO static pods by the new operator (through K8S API ,using MachineSet ? ) that will 'clear' the static pods.

Reply:
I'm not too concerned about the case where they're running at the same time. The new services should just crashloop on port conflicts until the old services are stopped, at which point the new services can come up and take over.

The timing is an issue though. It could break the cluster if the old services are deleted before the new ones have started.

Reply:
AFAIK, MCO upgrade executed independently in each node, so relying on port conflict can solve the problem in node scope and not cluster wide. we still need to address the case where some nodes run MCO stack and other nodes run the new operator stack.
Anyhow, I'll compile all the information and update the upgrade/downgrade section.